File: android/src/main/kotlin/com/google/flutter/recaptcha/RecaptchaEnterprisePlugin.kt
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.flutter.recaptcha

import android.app.Application
import androidx.annotation.NonNull
import com.google.android.recaptcha.Recaptcha
import com.google.android.recaptcha.RecaptchaAction
import com.google.android.recaptcha.RecaptchaClient
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

/** RecaptchaEnterprisePlugin */
class RecaptchaEnterprisePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
  private lateinit var channel: MethodChannel
  private lateinit var recaptchaClient: RecaptchaClient
  private lateinit var application: Application

  override fun onAttachedToEngine(
    @NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
  ) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "recaptcha_enterprise")
    channel.setMethodCallHandler(this)
  }

  fun mapAction(actionStr: String): RecaptchaAction {
    return when {
      actionStr.equals("login", ignoreCase = true) -> RecaptchaAction.LOGIN
      actionStr.equals("signup", ignoreCase = true) -> RecaptchaAction.SIGNUP
      else -> RecaptchaAction.custom(actionStr)
    }
  }

  private fun initClient(@NonNull call: MethodCall, @NonNull result: Result) {
    if (application == null) {
      result.error("FL_INIT_FAILED", "No application registered", null)
      return
    }
    val siteKey = call.argument<String>("siteKey")
    val timeout = call.argument<Long>("timeout")

    if (siteKey != null) {
      GlobalScope.launch {
        let {
            if (timeout != null) Recaptcha.getClient(application, siteKey, timeout)
            else Recaptcha.getClient(application, siteKey)
          }
          .onSuccess { client ->
            recaptchaClient = client
            result.success(true)
          }
          .onFailure { exception -> result.error("FL_INIT_FAILED", exception.toString(), null) }
      }
    }
  }

  private fun execute(@NonNull call: MethodCall, @NonNull result: Result) {
    if (!this::recaptchaClient.isInitialized || recaptchaClient == null) {
      result.error("FL_EXECUTE_FAILED", "Initialize client first", null)
      return
    }

    val actionStr = call.argument<String>("action")
    if (actionStr == null) {
      result.error("FL_EXECUTE_FAILED", "Missing action", null)
      return
    }

    val action = mapAction(actionStr)
    val timeout = call.argument<Long>("timeout")
    GlobalScope.launch {
      recaptchaClient
        .let { if (timeout != null) it.execute(action, timeout) else it.execute(action) }
        .onSuccess { token -> result.success(token) }
        .onFailure { exception -> result.error("FL_EXECUTE_FAILED", exception.toString(), null) }
    }
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "initClient" -> initClient(call, result)
      "execute" -> execute(call, result)
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    val activity = binding.activity
    application = activity.application
  }

  override fun onDetachedFromActivity() {}

  override fun onDetachedFromActivityForConfigChanges() {}

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}
}

File: ios/Sources/recaptcha_enterprise_flutter/RecaptchaEnterprisePlugin.swift
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Flutter
import RecaptchaEnterprise
import UIKit

public class RecaptchaEnterprisePlugin: NSObject, FlutterPlugin {
  var recaptchaClient: RecaptchaClient?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "recaptcha_enterprise", binaryMessenger: registrar.messenger())
    let instance = SwiftRecaptchaEnterprisePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  private func mapAction(_ actionStr: String) -> RecaptchaAction {
    if actionStr == "login" {
      return RecaptchaAction.login
    } else if actionStr == "signup" {
      return RecaptchaAction.signup
    } else {
      return RecaptchaAction(customAction: actionStr)
    }
  }

  private func createGetClientHandler(_ result: @escaping FlutterResult) -> (
    RecaptchaClient?, Error?
  ) -> Void {
    let getClientClosure: (RecaptchaClient?, Error?) -> Void = {
      recaptchaClient, error in
      if let recaptchaClient = recaptchaClient {
        self.recaptchaClient = recaptchaClient
        result(true)
      } else if let error = error {
        guard let error = error as? RecaptchaError else {
          FlutterError.init(
            code: "FL_CAST_ERROR", message: "Not a RecaptchaError", details: nil
          )
          return
        }
        result(
          FlutterError.init(
            code: String(error.code), message: error.errorMessage, details: nil)
        )
      }
    }
    return getClientClosure
  }

  private func fetchClient(
    _ call: FlutterMethodCall, result: @escaping FlutterResult
  ) {
    guard let args = call.arguments as? [String: Any],
      let siteKey = args["siteKey"] as? String
    else {
      result(
        FlutterError.init(
          code: "FL_INIT_FAILED", message: "Missing site key", details: nil))
      return
    }

    Recaptcha.fetchClient(
      withSiteKey: siteKey, completion: createGetClientHandler(result))
  }

  private func initClient(
    _ call: FlutterMethodCall, result: @escaping FlutterResult
  ) {
    guard let args = call.arguments as? [String: Any],
      let siteKey = args["siteKey"] as? String
    else {
      result(
        FlutterError.init(
          code: "FL_INIT_FAILED", message: "Missing site key", details: nil))
      return
    }

    let getClientClosure = createGetClientHandler(result)

    if let timeout = args["timeout"] as? Double {
      Recaptcha.getClient(
        withSiteKey: siteKey, withTimeout: timeout, completion: getClientClosure
      )
    } else {
      Recaptcha.getClient(withSiteKey: siteKey, completion: getClientClosure)
    }
  }

  private func execute(
    _ call: FlutterMethodCall, result: @escaping FlutterResult
  ) {
    guard let client = recaptchaClient else {
      result(
        FlutterError.init(
          code: "FL_EXECUTE_FAILED", message: "Initialize client first",
          details: nil))
      return
    }

    guard let args = call.arguments as? [String: Any],
      let actionStr = args["action"] as? String
    else {
      result(
        FlutterError.init(
          code: "FL_EXECUTE_FAILED", message: "Missing action", details: nil))
      return
    }

    let action = mapAction(actionStr)

    let executeClosure: (String?, Error?) -> Void = { token, error -> Void in
      if let token = token {
        result(token)
      } else if let error = error {
        guard let error = error as? RecaptchaError else {
          FlutterError.init(
            code: "FL_CAST_ERROR", message: "Not a RecaptchaError", details: nil
          )
          return
        }
        result(
          FlutterError.init(
            code: String(error.code), message: error.errorMessage, details: nil)
        )
      }
    }

    if let timeout = args["timeout"] as? Double {
      client.execute(
        withAction: action, withTimeout: timeout, completion: executeClosure)
    } else {
      client.execute(withAction: action, completion: executeClosure)
    }

  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult)
  {
    switch call.method {
    case "initClient":
      initClient(call, result: result)
    case "fetchClient":
      fetchClient(call, result: result)
    case "execute":
      execute(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
